### ISSUE TITLE
Some map<X>Array not working

### ISSUE THREAD

**Issue by momania** [bytefish/PgBulkInsert#20]
I try to use arrays, and get stuck with using numeric arrays and string arrays. Double arrays work as specified [in the test](https://github.com/bytefish/PgBulkInsert/blob/master/PgBulkInsert/src/test/java/de/bytefish/pgbulkinsert/pgsql/handlers/PgBulkInsertTest.java#L589), but numeric, or strings fails with the following error: `org.postgresql.util.PSQLException: ERROR: wrong element type`

Column definition is a simple: 
```
create table my_table (
    my_strings varchar[],
    my_numeric numeric[]
);
```

Rings a bell and am I overseeing something, or is this a missing feature? :-)

**Comment by bytefish** [comment]
Hmmm. No you are not missing something. 🤔 Can you write a Unit Test, that reproduces the bug? So I have a starting point for debugging. 💪

**Comment by bytefish** [comment]
So here is the first problem: The ``mapStringArray(...)`` method only maps to a PostgreSQL ``text`` column and not to a varchar:

![image](https://user-images.githubusercontent.com/239373/40442711-b773c7be-5ec4-11e8-9dee-ffc6624117a7.png)

Then I looked into the Object Identifier list of PgBulkInsert and saw... I didn't include a ``varchar`` at all. So yes, you have found a missing feature. 👍

**Comment by momania** [comment]
Ah great! I'll see if I can make the unit tests for the missing array mappings tomorrow, so you can make sure they all work 😄 
Now that you hinted that it's all in the data type definition mapping, I'll have a look if I can include the fix too.

**Comment by bytefish** [comment]
I am extending it right now and wrote a Unit Test. Give me a minute. 🥇

**Comment by bytefish** [comment]
OK. I fixed it:

* https://github.com/bytefish/PgBulkInsert/blob/master/PgBulkInsert/src/test/java/de/bytefish/pgbulkinsert/pgsql/handlers/ArrayTypesTest.java

But I won't manage to release it today. Also there is one more thing in the API Design, that needs to be adressed.

**Comment by bytefish** [comment]
The thing is. Everything works fine with the current API, if you say "mapStringArray" and your Column is of Type ``text``[]. Things get hairy, when let's say your field is a ``varchar[]`` like in your case. How do you want to differentiate between the Java Type and the Postgres Type. 

So at least for arrays, users need to pass in the specific types using ``mapCollection``. But that is not really intuitive and leads to errors.

I am thinking about renaming the ``mapStringArray`` into two methods:

* ``mapTextArray``
* ``mapVarCharArray``

Things get even more complicated with a numeric field. In Java you have BigDecimal to represent a Decimal with Precision and Scale. But... I am sure people have a ``Double`` / ``Double[]`` and want to map those to a numeric Field in the database.

Then what should the method look like? ``mapNumericFloat`` and ``mapNumericDouble``, ``mapNumericBigDecimal`` doesn't sound good to me. 👎 I am close to saying, there is a method ``mapNumeric``, which casts call all numbers to be written into a ``BigDecimal`` and calls the usual ``BigDecimal`` method.

What do you think?

**Comment by momania** [comment]
I agree that people (should) know their databases and you don't have to make this a magic layer that knows how to magically translate java types to database types, so it makes more sense to use the database types in the method names, and possibly, where applicable, use method overloading for the various compatible java types.

So for the numeric case I would add the `mapNumericArray(Array<BigDecimal>)` for sure, and maybe add convenience method `mapNumericArray(Array<Double>)` for instance.

Do you want me to try and create a pull request for it, or do you want to continue it yourself?

**Comment by bytefish** [comment]
I have added an initial version of how a mapping for a int, long, float, double properties to numeric could look like. You cannot overload generics in Java, so you have to pass an additional type token into the method (welcome Type Erasure). It's not a beauty, but a somewhat obvious and straightforward mapping. 

I am still thinking about what an API for Arrays could look like. 🤔 

My ``mapCollection`` implementation for resolving the handler by its Postgres type is simply wrong, if you want to map a double array to a numeric column. It will load the wrong default handler (for a BigDecimal) and it either crashes or produces wrong results - both outcomes are not optimal. 😌

If you have any suggestions for a better API or how to restructure the internals, let me know.

**Comment by momania** [comment]
Have a look at my commit mentioned above here. or click [here](https://github.com/momania/PgBulkInsert/commit/fa1cc7afc4fbdf0dc299545ca2c7aebadb96d889)

I think this is a setup that can work. Just need to add more unit tests so more array types get some coverage.

If you're happy with it I can make a pull request for it.



I also found some weird behaviour with inserting numeric arrays on my side, but haven't looked into that yet. But it looked like all inserted BigDecimals were wrong.

**Comment by bytefish** [comment]
Like Aretha Franklin said: Respect! I will take a look, when I get home.

**Comment by bytefish** [comment]
Wow you have really put a lot of effort into it! Looks good! Would you mind making a Pull Request, so we can iterate on it?

**Comment by momania** [comment]
You're welcome. Was not too much effort though. Good use of IntelliJ helps a lot with refactoring. 😄 

But I like the lib and it's very helpful to a project I work on, so I might as well get my hands dirty if thing can be improved. 👍

**Comment by bytefish** [comment]
@momania I have added a unit test for the ``numeric`` mappings. I think it is ok to cast the numeric to a ``double``, when the type is not a ``BigDecimal``. :+1: While at it I also added you to the List of authors in the ``pom.xml``. 

Thanks again for your kind help!

I have released version 2.2 to Maven, so it should be available in some hours. 

I am closing the ticket, feel free to reopen, if there are problems.

**Comment by momania** [comment]
👍
